File: frontend/src/lib/preset-contracts.ts
/** Attach-preset traits and lead-scoped staffing — omitted fields use defaults. */

import { z } from "zod";

export const PresetTraitFieldsSchema = z.object({
  reads_images: z.boolean().optional().default(false),
  can_manage: z.boolean().optional().default(false),
  implements: z.boolean().optional().default(true),
});

export const PresetSchema = z
  .object({
    id: z.string(),
    title: z.string(),
    name: z.string(),
    adapter: z.string(),
    command: z.string(),
    created_at: z.number(),
  })
  .extend(PresetTraitFieldsSchema.shape);
export type Preset = z.infer<typeof PresetSchema>;

export const StaffingTraitsSchema = z.object({
  reads_images: z.boolean(),
  can_manage: z.boolean(),
  implements: z.boolean(),
});

export const StaffingPresetSchema = z
  .object({
    id: z.string(),
    title: z.string(),
    name: z.string(),
    adapter: z.string(),
  })
  .extend(StaffingTraitsSchema.shape);

export const StaffingProcessSchema = z.object({
  line_id: z.string(),
  line: z.string(),
  handle: z.string(),
  adapter: z.string(),
  captain: z.boolean(),
  matched_preset: z.object({ id: z.string(), name: z.string(), title: z.string() }).nullable(),
  traits: StaffingTraitsSchema.nullable(),
});

export const StaffingSchema = z.object({
  presets_in_use: z.boolean(),
  presets: z.array(StaffingPresetSchema),
  processes: z.array(StaffingProcessSchema),
});
export type Staffing = z.infer<typeof StaffingSchema>;

File: partyline/preset_contracts.py
"""Wire contracts for attach presets and lead-scoped staffing."""

from pydantic import BaseModel, Field


class PresetIn(BaseModel):
    title: str
    name: str
    adapter: str = "opencode"
    command: str = ""
    reads_images: bool = False
    can_manage: bool = False
    implements: bool = True


class PresetResponse(BaseModel):
    id: str
    title: str
    name: str
    adapter: str
    command: str
    created_at: float
    reads_images: bool = False
    can_manage: bool = False
    implements: bool = True


class StaffingTraits(BaseModel):
    reads_images: bool
    can_manage: bool
    implements: bool


class StaffingPreset(StaffingTraits):
    id: str
    title: str
    name: str
    adapter: str


class MatchedPreset(BaseModel):
    id: str
    name: str
    title: str


class StaffingProcess(BaseModel):
    line_id: str
    line: str
    handle: str
    adapter: str
    captain: bool
    matched_preset: MatchedPreset | None = None
    traits: StaffingTraits | None = None


class StaffingLine(BaseModel):
    id: str
    name: str
    accepted_sha: str


class StaffingResponse(BaseModel):
    presets_in_use: bool
    presets: list[StaffingPreset] = Field(default_factory=list)
    processes: list[StaffingProcess] = Field(default_factory=list)
    lines: list[StaffingLine] = Field(default_factory=list)
